### PR TITLE
Update flat_hash_map

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14819 Implement c10::Tensor&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13348039/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14820 Convert caffe2/aten Tensors to/from c10&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13348044/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15195 Use C10Tensor in the dispatcher&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13461640/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15324 Fix C10_API/C10_EXPORT for op schema registration&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13500679/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#15367 Update flat_hash_map**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13510912/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15199 Move LayerNorm op schema to c10&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13462124/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15243 Enable calling caffe2 LayerNorm from PyTorch and JIT&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13473791/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15312 Remove Context from c10 operator schemas&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13495323/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15316 Move files to/from c10/core and c10/util&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13498493/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15317 Clean up Half&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13498492/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15407 caffe2::Tensor::is_same()&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13520389/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15853 Move blob to c10&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605256/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15854 Code style cleanup&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605258/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15855 Remove some dependencies from ivalue.h to ATen&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605259/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15856 Fix export macros&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605257/)

This updates flat_hash_map and fixes an issue with singletons across library boundaries
(see the PRs linked at the top of the file)

Differential Revision: [D13510912](https://our.internmc.facebook.com/intern/diff/D13510912/)

### CODE REVIEW

File: c10/util/flat_hash_map.h [hunk 1]
```diff
@@ -1,5 +1,11 @@
-// Taken from https://github.com/skarupke/flat_hash_map/blob/2ceada9f7fb5c6d808e33546882fc63662e10e79/flat_hash_map.hpp
-//
+// Taken from https://github.com/skarupke/flat_hash_map/blob/2c4687431f978f02a3780e24b8b701d22aa32d9c/flat_hash_map.hpp
+// with fixes applied:
+// - https://github.com/skarupke/flat_hash_map/pull/8
```
Comment (ezyang) [inline c10/util/flat_hash_map.h]: Why wasn't this problem caught by warnings? It's the stereotypical `-Wall` thing a compiler should tell you about.

**Comment by ezyang** [comment]
I can't approve this diff because I don't understand the upstream changes.

**Comment by smessmer** [comment]
There's 4 upstream commits since the version we had checked in:
- https://github.com/skarupke/flat_hash_map/commit/b02e03371a33d9d4f50b82f98900175b33a64fd3 : Reduce constexpr trickery to make it work on more compilers (this actually makes some of the fixes obsolete I had built into our copy before)
- https://github.com/skarupke/flat_hash_map/commit/1c996c925ad9b6ce1f16906e9458e37491087c1d : Another fix for MSVC
- https://github.com/skarupke/flat_hash_map/commit/812aede752d789033a7c439e8fd4f8d81522a642 : Change default mapping from hash value to table indices to use fibonacci hashing
- https://github.com/skarupke/flat_hash_map/commit/2c4687431f978f02a3780e24b8b701d22aa32d9c : Fix that fibonacci hashing for size 1 hash tables.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
These bug fixes make me nervous about the robustness about this hash map implementation, compared to other choices we could have made.

**Comment by smessmer** [comment]
Yes, I got a bit wary too. I see two options - (1) add extensive test cases or (2) use another implementation, for example the google one.

**Pull request closed without merging**